cmd/akamai-purger/main.go

```diff
@@ -24,40 +23,150 @@ import (
+	// responses, assuming 3 URLs per request, is about 6 hours of work using
+	// default settings.
+	defaultQueueSize         = 1250000
+	defaultResponsesPerBatch = 2
+	defaultPurgeBatchEvery   = time.Millisecond * 32
+
+	// akamaiBytesPerResponse is the total bytes of all 3 URLs of cached OCSP responses associated with
+	// each certificate.
+	akamaiBytesPerResponse = 400
+
+	// akamaiBytesPerReqLimit is the limit of bytes allowed in a single request
+	// to the Fast-Purge API. With a limit of no more than 50,000 bytes, we
+	// subtract 1 byte to get the limit, and subtract an additional 19 bytes for
+	// overhead of the 'objects' key and array.
+	akamaiBytesPerReqLimit = 50000 - 1 - 19
+
+	// akamaiAPIReqPerSecondLimit is the limit of requests, per second, that
+	// we're allowed to make to the Fast-Purge API.
+	akamaiAPIReqPerSecondLimit = 50
+
+	// akamaiAPIReqPerSecondLimit is the limit of URLs, sent per second, that
+	// we're allowed to make to the Fast-Purge API.
+	akamaiURLsPerSecondLimit = 200
+)
+
+// Throughput is a container for all throuput related akamai-purger
+// configuration settings.
+type Throughput struct {
+
+	// ResponsesPerBatch the number of cached OCSP responses to included in each
+	// purge request. One cached OCSP response is composed of 3 URLs totaling <
+	// 400 bytes. If this value isn't provided it will default to
+	// 'defaultResponsesPerBatch'.
+	ResponsesPerBatch int
+
+	// PurgeBatchEvery is the duration waited between dispatching an Akamai
+	// purge request containing 'ResponsesPerBatch' * 3 URLs. If this value
+	// isn't provided it will default to 'defaultPurgeBatchEvery'.
+	PurgeBatchEvery cmd.ConfigDuration
+}
+
+func (t *Throughput) useOptimizedDefaults() {
+	if t.ResponsesPerBatch == 0 {
+		t.ResponsesPerBatch = defaultResponsesPerBatch
+	}
+	if t.PurgeBatchEvery.Duration == 0 {
+		t.PurgeBatchEvery.Duration = defaultPurgeBatchEvery
+	}
+}
+
+// validate ensures that the provided throughput configuration will not violate
+// the Akamai Fast-Purge API limits. For more information see the official
+// documentation:
+// https://techdocs.akamai.com/purge-cache/reference/rate-limiting
+func (t *Throughput) validate() error {
+	if t.PurgeBatchEvery.Duration == 0 {
+		// TODO(#6003) remove /'purgeInterval'.
+		return errors.New("'purgeBatchEvery'/'purgeInterval' must be > 0 nanoseconds")
+	}
+	if t.ResponsesPerBatch <= 0 {
+		return errors.New("'responsesPerBatch' must be > 0")
+	}
+
+	// Send no more than the 50,000 bytes of objects we’re allotted per request.
+	bytesPerRequest := (t.ResponsesPerBatch * akamaiBytesPerResponse)
+	if bytesPerRequest > akamaiBytesPerReqLimit {
+		return fmt.Errorf("configuration exceeds Akamai's bytes per request limit (%d bytes) by %d",
+			akamaiBytesPerReqLimit, bytesPerRequest-akamaiBytesPerReqLimit)
+	}
+
+	// Send no more than the 50 API requests we’re allotted each second.
+	requestsPerSecond := int(time.Second / t.PurgeBatchEvery.Duration)
```

Comment:
the `int()` conversion is going to truncate downwards, which means that certain configurations could result in us attempting to send _nearly_ 51 requests per second.

Reply:
I've fixed the code and modified my requests-per-second test to reflect this case. FWIW this is a duration that can only be expressed in microseconds and smaller, still it's correct.